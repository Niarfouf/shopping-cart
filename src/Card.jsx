export default function Card({ title, price, description, imageUrl }) {
    return (
        <>
            <div className="card">
                <img className="item-image" alt={title} src={imageUrl}></img>
                <h3 className="title">{title}</h3>
                <p>{description}</p>
                <input type="number"></input>
                <button>Add</button>
                <p>{price}</p>
            </div>
        </>
    )
}
