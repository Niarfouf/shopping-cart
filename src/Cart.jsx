export default function Cart({ cartinfo }) {
    return <></>
}
