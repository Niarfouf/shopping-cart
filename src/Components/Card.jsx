import { useEffect, useState } from 'react'
export default function Card({ item, addItemInList, quantity = 0, action }) {
    const [input, setInput] = useState(0)
    function left() {
        setInput((input) => (input > 0 ? input - 1 : input))
    }
    function right() {
        setInput((input) => input + 1)
    }
    useEffect(() => {
        setInput(quantity)
    }, [quantity])
    return (
        <>
            <div className="card">
                <img
                    className="item-image"
                    alt={item.title}
                    src={item.image}
                ></img>
                <h3 className="title">{item.title}</h3>

                <div>
                    <div className="add-container">
                        <label htmlFor={item.id}>Quantity :</label>
                        <button onClick={left} className="left-arrow-input">
                            <svg
                                xmlns="http://www.w3.org/2000/svg"
                                viewBox="0 0 24 24"
                            >
                                <title>menu-left</title>
                                <path d="M14,7L9,12L14,17V7Z" />
                            </svg>
                        </button>
                        <input
                            value={input}
                            onChange={(e) => setInput(e.target.value * 1)}
                            type="text"
                            inputMode="numeric"
                            name="quantity"
                            id={item.id}
                        ></input>
                        <button onClick={right} className="right-arrow-input">
                            <svg
                                xmlns="http://www.w3.org/2000/svg"
                                viewBox="0 0 24 24"
                            >
                                <title>menu-right</title>
                                <path d="M10,17L15,12L10,7V17Z" />
                            </svg>
                        </button>
                        <button
                            className="add-button"
                            onClick={() => addItemInList(item, input)}
                        >
                            {action}
                        </button>
                    </div>
                    <p className="price">
                        {input} unit x {item.price}€ ={' '}
                        {Math.round(item.price * input * 100) / 100} €
                    </p>
                </div>
            </div>
        </>
    )
}
