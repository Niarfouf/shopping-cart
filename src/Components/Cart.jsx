import Card from './Card'
export default function Cart({
    cartItems,

    deleteCartItems,
}) {
    return (
        <>
            <div className="shop">
                <h2>Cart</h2>
                <div className="item-cards">
                    {Object.entries(cartItems).map((item) => {
                        return (
                            <Card
                                key={item[0]}
                                item={item[1].item}
                                addItemInList={deleteCartItems}
                                quantity={item[1].quantity}
                                action={'Update'}
                            ></Card>
                        )
                    })}
                </div>
            </div>
            <div className="total-cart">
                <p className="total-price">
                    Total price ={' '}
                    {Object.entries(cartItems).reduce(
                        (acc, current) =>
                            acc +
                            Math.round(
                                current[1].quantity *
                                    current[1].item.price *
                                    100
                            ) /
                                100,
                        0
                    )}{' '}
                    â‚¬
                </p>
                <button className="confirm-cart">Confirm</button>
            </div>
        </>
    )
}
