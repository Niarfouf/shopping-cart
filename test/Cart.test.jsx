import { vi, beforeAll, afterAll } from 'vitest'
import { render, act, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Cart from '../src/Components/Cart'
import Card from '../src/Components/Card'
import { MemoryRouter } from 'react-router-dom'

const item1 = {
    title: 'testTitle1',
    image: 'https://fakestoreapi.com/img/61U7T1koQqL._AC_SX679_.jpg',
    id: 0,
    price: 5.5,
}
const item2 = {
    title: 'testTitle2',
    image: 'https://fakestoreapi.com/img/71kWymZ+c+L._AC_SX679_.jpg',
    id: 1,
    price: 10,
}
const cart = {
    0: { item: item1, quantity: 1 },
    1: { item: item2, quantity: 2 },
}
const cart2 = {
    0: { item: item1, quantity: 5 },
    1: { item: item2, quantity: 2 },
}

describe('Cart', () => {
    it('render snapshot', () => {
        const { container } = render(
            <Cart cartItems={cart} deleteCartItems={() => {}} />
        )
        expect(container).toMatchSnapshot()
    })
    it('render correct header', () => {
        render(<Cart cartItems={cart} deleteCartItems={() => {}} />)

        expect(screen.getAllByRole('heading')[0].textContent).toMatch(/Cart/i)
    })
    it('render correct number of items', () => {
        render(<Cart cartItems={cart} deleteCartItems={() => {}} />)

        expect(screen.getAllByRole('heading').length).toBe(3)
    })
    it('render correct total price', () => {
        render(<Cart cartItems={cart} deleteCartItems={() => {}} />)

        expect(screen.getByText('Total price = 25.5 â‚¬')).toBeInTheDocument()
    })
})
