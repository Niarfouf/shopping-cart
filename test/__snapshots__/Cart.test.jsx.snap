// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart > render snapshot 1`] = `
<div>
  <div
    class="shop"
  >
    <h2>
      Cart
    </h2>
    <div
      class="item-cards"
    >
      <div
        class="card"
      >
        <img
          alt="testTitle1"
          class="item-image"
          src="https://fakestoreapi.com/img/61U7T1koQqL._AC_SX679_.jpg"
        />
        <h3
          class="title"
        >
          testTitle1
        </h3>
        <div>
          <div
            class="add-container"
          >
            <label
              for="0"
            >
              Quantity :
            </label>
            <button
              class="left-arrow-input"
            >
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  menu-left
                </title>
                <path
                  d="M14,7L9,12L14,17V7Z"
                />
              </svg>
            </button>
            <input
              id="0"
              inputmode="numeric"
              name="quantity"
              type="text"
              value="1"
            />
            <button
              class="right-arrow-input"
            >
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  menu-right
                </title>
                <path
                  d="M10,17L15,12L10,7V17Z"
                />
              </svg>
            </button>
            <button
              class="add-button"
            >
              Update
            </button>
          </div>
          <p
            class="price"
          >
            1
             unit x 
            5.5
            € =
             
            5.5
             €
          </p>
        </div>
      </div>
      <div
        class="card"
      >
        <img
          alt="testTitle2"
          class="item-image"
          src="https://fakestoreapi.com/img/71kWymZ+c+L._AC_SX679_.jpg"
        />
        <h3
          class="title"
        >
          testTitle2
        </h3>
        <div>
          <div
            class="add-container"
          >
            <label
              for="1"
            >
              Quantity :
            </label>
            <button
              class="left-arrow-input"
            >
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  menu-left
                </title>
                <path
                  d="M14,7L9,12L14,17V7Z"
                />
              </svg>
            </button>
            <input
              id="1"
              inputmode="numeric"
              name="quantity"
              type="text"
              value="2"
            />
            <button
              class="right-arrow-input"
            >
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  menu-right
                </title>
                <path
                  d="M10,17L15,12L10,7V17Z"
                />
              </svg>
            </button>
            <button
              class="add-button"
            >
              Update
            </button>
          </div>
          <p
            class="price"
          >
            2
             unit x 
            10
            € =
             
            20
             €
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    class="total-cart"
  >
    <p
      class="total-price"
    >
      Total price =
       
      25.5
       
      €
    </p>
    <button
      class="confirm-cart"
    >
      Confirm
    </button>
  </div>
</div>
`;
